File: spring-integration/basic-integration/src/main/java/com/ddubson/integration/IntegrationConfig.java
package com.ddubson.integration;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.integration.channel.DirectChannel;
import org.springframework.integration.dsl.IntegrationFlow;
import org.springframework.integration.dsl.IntegrationFlows;

@Configuration
public class IntegrationConfig {
    private final PrintService printService;

    public IntegrationConfig(PrintService printService) {
        this.printService = printService;
    }

    @Bean("inputChannel")
    public DirectChannel inputChannel() {
        return new DirectChannel();
    }

    @Bean("outputChannel")
    public DirectChannel outputChannel() {
        return new DirectChannel();
    }

    /**
     * Input Channel
     * -> Print Service (Service Activator)
     * -> Output Channel
     */
    @Bean
    public IntegrationFlow basicFlow() {
        return IntegrationFlows
                .from(inputChannel())
                .handle(printService)
                .channel(outputChannel())
                .get();
    }
}

File: spring-integration/basic-integration/src/main/kotlin/com/ddubson/integration/IntegrationConfig.kt
package com.ddubson.integration

import com.ddubson.CommandLineLogAdapter
import com.ddubson.LogAdapter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.integration.annotation.ServiceActivator
import org.springframework.integration.channel.DirectChannel
import org.springframework.integration.dsl.IntegrationFlow
import org.springframework.integration.dsl.IntegrationFlows
import org.springframework.integration.dsl.channel.MessageChannels

@Configuration
class IntegrationConfig {
    /**
     * Input Channel
     * -> Print Service (Service Activator)
     * -> Output Channel
     */
    //@Bean
    /*fun basicFlow(): IntegrationFlow {
        return IntegrationFlows
                .from(inputChannel())
                .handle(printService())
                .channel(outputChannel())
                .get()
    }*/

    @Bean("inputChannel")
    fun inputChannel(): DirectChannel = MessageChannels.direct().get()

    @Bean("outputChannel")
    fun outputChannel(): DirectChannel = MessageChannels.direct().get()

    @Bean
    fun logAdapter(): LogAdapter = CommandLineLogAdapter()

    @Bean
    @ServiceActivator(inputChannel = "inputChannel", outputChannel = "outputChannel")
    fun printService(): PrintService = PrintService(logAdapter())
}